### PR TITLE
union: support occupying horizontal multiple fields for list field

### LINKED ISSUE #171: union: if the last field of a union is an incell list, extend it just like a horizontal list
A union named `Entity` is defined below for example:
| Name | Alias | Field1 | Field2 | Field3 |
|--|--|--|--|--|
| Hero |  | Name <br> string | Level <br> uint32 | Abilities <br> []enum<.Ability>\|{extend:true} |
| Monster |  | Name <br> string | Awards<br> [].Item\|{extend:true} |  |

We mark `Hero.Abilities` `entend:true`, so we can use Field3 and all following fields as elements of this list, one cell for each element just like a horizontal list. `Monster.Awards` is also similar, but note that we should use sep `,` rather than subsep `:` to seperate between `item_id` and `item_num`.

Use this union in a sheet named `EntityConf` for example:
| EntityType | EntityField1 | EntityField2 | EntityField3 | EntityField4 | EntityField5 | 
|--|--|--|--|--|--|
| Hero  | Berserker | 60 | ABILITY_HASTE | ABILITY_CRIT | |
| Hero  | Paladin | 80 | ABILITY_SHIELD | ABILITY_HEAL | ABILITY_CRIT  |
| Monster  | Goblin | 1001,100 | 1002,50 |  |  |

### PR DESCRIPTION
**Pull request by Kybxd**
* If a union field has property: `cross:x`(x not zero), it occupies x fields of this union. For example, if a union's Field2 has type `[]int64|{cross:3}`, it occupies 3 fields (Field2, Field3, Field4), and if a next field exists, it should be defined at Field5.
* If a union field has property: `cross:-1`, it occupies all following fields of this union. No next field should exist after a `cross:-1` field.

- [x] close #171 
- [x] table parser testing
- [x] confgen table parser: encaplulate standalone functions to parse map/list at different layouts for code reuse
- protogen: document parser not supported

### CODE REVIEW

File: proto/tableau/protobuf/tableau.proto [hunk 1]
```diff
@@ -253,6 +253,10 @@ message FieldProp {
   bool optional = 11;
   // Field patch type.
   Patch patch = 12;
+  // Entend the last field of a union if it's an incell list. This list now
+  // takes up the current cell and all following cells in the same union, each
+  // cell one element just like a horizontal list.
+  bool extend = 13;
```
Comment (wenchy) [inline proto/tableau/protobuf/tableau.proto:259]: Just use `size` option (it is designed for this alike condition), no need to add extra `extend` option.

**Review by wenchy — CHANGES_REQUESTED** [review]

File: test/functest/proto/excel__union__union.proto [hunk 2]
```diff
@@ -31,5 +32,6 @@ message PredefinedUnion {
     int32 id = 1 [(tableau.field) = {name:"ID"}];
     union.Target target = 2 [(tableau.field) = {name:"Target"}];
     int32 progress = 3 [(tableau.field) = {name:"Progress"}];
+    protoconf.BattleTarget battle_target = 4 [(tableau.field) = {name:"BattleTarget"}];
```
Comment (wenchy) [inline test/functest/proto/excel__union__union.proto:35]: 1. DO NOT mix with the `PredefinedUnion` sheet, just add a new sheet `UnionFieldsOption` to test `union_fields` feature. And should add new union type sheet to define different union fields type.
2. `UnionFieldsOption` should test following cases:
    - scalar list
    - enum list
    - struct list
 3. Also should test not zero value, like `union_fields: 3` case

File: internal/confgen/document_parser.go [hunk 3]
```diff
@@ -527,13 +527,37 @@ func (sp *documentParser) parseUnionMessage(field *Field, msg protoreflect.Messa
 					)
 					return xerrors.WrapKV(xerrors.E2014(subField.opts.Name), kvs...)
 				}
-				err = sp.parser.parseUnionMessageField(subField, msg, valNode.Value)
+				var nodeDatas []string
```
Comment (wenchy) [inline internal/confgen/document_parser.go:530]: naming: `nodeDatas` -> `nodeDataList` or `nodeValues`

NOTE: `data` is an uncountable noun, so there is no `datas` plural word!

**Review by wenchy — CHANGES_REQUESTED** [review]

File: proto/tableau/protobuf/tableau.proto (hunk 1 above)
Comment (Kybxd) [reply proto/tableau/protobuf/tableau.proto]: done

File: test/functest/proto/excel__union__union.proto (hunk 2 above)
Comment (Kybxd) [reply test/functest/proto/excel__union__union.proto]: done

File: internal/confgen/document_parser.go [hunk 4]
```diff
@@ -527,13 +527,37 @@ func (sp *documentParser) parseUnionMessage(field *Field, msg protoreflect.Messa
...
+						}
+						nodeDatas = append(nodeDatas, node.Value)
+					}
+				default:
+					nodeDatas = []string{valNode.Value}
+					for j := 1; j < int(subField.opts.Prop.GetUnionFields()); j++ {
+						nodeName := unionDesc.ValueFieldName() + strconv.Itoa(int(fd.Number())+j)
+						node := node.FindChild(nodeName)
+						if node == nil {
+							break
```
[the remark this answers is not in the document]
Comment (Kybxd) [reply internal/confgen/document_parser.go]: done

File: internal/confgen/parser.go [hunk 5]
```diff
@@ -1331,12 +1331,53 @@ func (sp *sheetParser) parseUnionMessageField(field *Field, msg protoreflect.Mes
...
 		// incell list
 		value := msg.NewField(field.fd)
-		present, err := sp.parseIncellListField(field, value.List(), cellData)
-		if err != nil {
-			return err
-		}
-		if present {
-			msg.Set(field.fd, value)
+		list := value.List()
+		if len(cellDatas) != 0 {
```
[the remark this answers is not in the document]
Comment (Kybxd) [reply internal/confgen/parser.go]: done

File: proto/tableau/protobuf/tableau.proto [hunk 6]
```diff
@@ -253,6 +253,11 @@ message FieldProp {
   bool optional = 11;
   // Field patch type.
   Patch patch = 12;
+  // Modify the layout of a list field of union's nested message 
+  // from incell to horizontal, and specify the field num 
+  // the list now occupies(one list element for each field). 
+  // Zero value means the list occupies all following fields.
+  optional uint32 union_fields = 13;
```
[the remark this answers is not in the document]
Comment (wenchy) [reply proto/tableau/protobuf/tableau.proto]: done

**Review by wenchy — CHANGES_REQUESTED** [review]

**Review by wenchy — APPROVED** [review]
